events/bingo/webApps/April 04 Bingo Tool/april04bingoWinners.js

```diff
@@ -0,0 +1,251 @@
+//
+// april04bingoWinners.js
+// NodeJS Web App for Bingo to fix some mistakes
+// Created by Zach Fox on 2019-04-04
+// Copyright High Fidelity 2019
+//
+// Licensed under the Apache 2.0 License
+// See accompanying license file or http://apache.org/
+//
+
+function checkRow(userNumbers2D, calledNumbers, rowIndex) {
+    for (var colIterator = 0; colIterator < NUM_ROWS_COLS; colIterator++) {
+        if (calledNumbers.indexOf(userNumbers2D[colIterator][rowIndex]) === -1) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function checkColumn(userNumbers2D, calledNumbers, columnIndex) {
+    for (var rowIterator = 0; rowIterator < NUM_ROWS_COLS; rowIterator++) {
+        if (calledNumbers.indexOf(userNumbers2D[columnIndex][rowIterator]) === -1) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function checkDiagonalTopLeftToBottomRight(userNumbers2D, calledNumbers) {
+    for (var rowIterator = 0; rowIterator < NUM_ROWS_COLS; rowIterator++) {
+        if (calledNumbers.indexOf(userNumbers2D[rowIterator][rowIterator]) === -1) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function checkDiagonalTopRightToBottomLeft(userNumbers2D, calledNumbers) {
+    for (var rowIterator = 0; rowIterator < NUM_ROWS_COLS; rowIterator++) {
+        // eslint-disable-next-line no-magic-numbers
+        if (calledNumbers.indexOf(userNumbers2D[4 - rowIterator][rowIterator]) === -1) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+var NUM_ROWS_COLS = 5;
+var CENTER_ROW_COL_IDX = 2;
+function verifyBingoWin(userNumbers, calledNumbers) {
+    // -1 is the free space and was always "called"
+    calledNumbers.push(-1);
+
+    var userNumbers2D = [];
+    for (var rowIterator = 0; rowIterator < NUM_ROWS_COLS; rowIterator++) {
+        var currentColumn = [];
+
+        for (var colIterator = 0; colIterator < NUM_ROWS_COLS; colIterator++) {
+            // Handle free space
+            if (rowIterator === CENTER_ROW_COL_IDX && colIterator === CENTER_ROW_COL_IDX) {
+                currentColumn.push(-1);
+            } else {
+                currentColumn.push(userNumbers.shift());
+            }
+        }
+        userNumbers2D.push(currentColumn);
+    }
+
+    for (var i = 0; i < NUM_ROWS_COLS; i++) {
+        if (checkRow(userNumbers2D, calledNumbers,i)) {
+            return true;
+        }
+
+        if (checkColumn(userNumbers2D, calledNumbers, i)) {
+            return true;
+        }
+    }
+
+    if (checkDiagonalTopLeftToBottomRight(userNumbers2D, calledNumbers) || checkDiagonalTopRightToBottomLeft(userNumbers2D, calledNumbers)) {
+        return true;
+    }
+    
+    return false;
+}
+
+
+function getAllLogicalWinners(oneTableData) {
+    var bossNumbers = [];
+    for (var i = 0; i < oneTableData.length; i++) {
+        if (oneTableData[i].username === "BINGO BOSS") {
+            bossNumbers = JSON.parse(oneTableData[i].cardNumbers);

```

Comment:
Try catch for parsing!

Reply:
not necessary because i know for certain the format of the data